beacon-chain/blockchain/process_block.go

```diff
@@ -648,10 +643,6 @@ func (s *Service) lateBlockTasks(ctx context.Context) {
 	}
 
 	attribute := s.getPayloadAttribute(ctx, headState, s.CurrentSlot()+1, headRoot[:])
-	// return early if we are not proposing next slot
-	if attribute.IsEmpty() {
-		return
-	}
 
 	s.headLock.RLock()
```

Comment:
This changes an invariant: previously, we didn’t need to call the head block if we weren’t proposing the next block, as calling the head block isn’t free. Now, we must call the head block even if we’re not proposing the next block when a late block is triggered. This might be fine, but if we want to be precise about it, we could move the head block call inside the goroutine

Reply:
I've changed the event firing code and the handler to accommodate the head block being nil, looking up the block in the handler if it is nil, and making sure the root matches the head root in the event. If the head root in the event is different, we just drop the event. In this circumstance we know the mismatched event was fired by `lateBlockTasks`, then essentially preempted by block processing updated head. The head update will stream its own proposer attribute event which is what the consumer really wants, so it should be fine to drop the late block event in this race.